crates/sqlbuiltins/src/functions/scalars/mod.rs

```diff
@@ -51,16 +54,127 @@ impl BuiltinScalarUDF for Version {
     }
 }
 
-fn get_nth_scalar_value(input: &[ColumnarValue], n: usize) -> Option<ScalarValue> {
+fn get_nth_scalar_value(
+    input: &[ColumnarValue],
+    n: usize,
+    op: &dyn Fn(ScalarValue) -> Result<ScalarValue, BuiltinError>,
+) -> Result<ColumnarValue, BuiltinError> {
     match input.get(n) {
         Some(input) => match input {
-            ColumnarValue::Scalar(scalar) => Some(scalar.clone()),
-            ColumnarValue::Array(arr) => ScalarValue::try_from_array(arr, 0).ok(),
+            ColumnarValue::Scalar(scalar) => Ok(ColumnarValue::Scalar(op(scalar.clone())?)),
+            ColumnarValue::Array(arr) => {
+                let mut values = Vec::with_capacity(arr.len());
```

Comment:
Generally, datafusion favors operating on the arrays instead of the scalars, and even has a helper function [make_scalar_function](https://github.com/apache/arrow-datafusion/blob/28ca6d1ad9692d0f159ed1f1f45a20c0998a47ea/datafusion/physical-expr/src/functions.rs#L196-L197) to do so. 

a crude (non working) example of how i think we should approach this instead. 
```rust
 let func =  (|args: &[ArrayRef]| {
            let filter = args.get(1);
            let filter = filter.and_then(|arg| match arg.as_any().downcast_ref::<StringArray>() {
                Some(arr) => Some(arr.value(0)),
                None => match arg.as_any().downcast_ref::<LargeStringArray>() {
                    Some(arr) => Some(arr.value(0)),
                    None => None,
                },
            }).expect("invalid type for KQL expression");

            let docs = match args.get(0).and_then(|docs| {
                match docs.data_type() {
                    DataType::Utf8 => docs.as_any().downcast_ref::<StringArray>().unwrap().into_iter().map(|s| {
                        kdl_inner_logic(s, filter)
                    }).collect(),
                    DataType::LargeUtf8 => docs.as_any().downcast_ref::<LargeStringArray>().unwrap().into_iter().map(|s| {
                        kdl_inner_logic(s, filter)
                    }).collect(),
                    _ => None,
                }
            })
        });

        let udf = ScalarUDF {
            // ...
            fun: func,
        };
```

it's a bit more boilerplate for the implementor of the udfs, but it better aligns with the datafusion udf examples, and provides more flexibility.

Reply:
To be clear, we're only ever calling `op` once per value (`O(n)`), we are converting the array to scalars and storing them in a vector before calling the `op` on them, so the memory use is: [(original data) + (result set) + vector for the working space] (which is handled in batches so less than `n`), and the alternative is that we could avoid the working space vector, though some of this depends on how big the arrays are: the cases where it's actually 2n means that the total data size is small, and therefore less of a problem.

The alternative is possible, but it also comes with some cost/trade offs:
- `ScalarValue::iter_to_array` does not take a result, so while we could implement this without the vector as: `(0..arr.len()).into_iter().map(|idx| ScalarValue::try_from_array(arr, idx).and_then(op))`, you end up ignoring errors either in converting values (until the end) or in the operation itself. I felt like the extra memory was worth the error handling simplicity.
- We could implement (or modify) a `ScalarValue::iter_to_aray` variant that handled/aborted on errors. This is probably a good idea, generally, but I think we should do it in the upstream rather than in our own code, and in the mean time.
- handling columns of arbitrary types is hard, and so if we don't allow ourselves this kind of operation, we'll end up having functions that are limited to only some types (which is more friction for users), or handling the same type conversions in different places. 
- if we don't allow ourselves to convert columns to scalars in functions writing functions for GlareDB (of any kind) becomes much more difficult and our time to delivery goes down. 

Regardless of how we handle the processing/loop if a function is written to handle scalar data we have to convert the data to scalar at some point, and I think saying "all UDFs should be zero-copy" shouldn't be our goal (certainly now; we can definitely optimize later), and I think we can optimize these later, as needed. 

> if the udfs operate on the column instead of the rows, it allows them more control to do optimizations that may only be applicable to their specific case.

Sure. This is a helper function that is non-compulsory, and I don't think any of the cases where it's being used would be particularly impacted, though implementers should consider this on a case by-case basis.

Reply:
(n.b. I'm pretty sure your implementation has the same memory properties as mine, though it's not particularly easy to read so I might be missing something. I think the extra boilerplate is very much something that we should avoid).

I think your example is an interesting start, but I'd really like to see a proposal that allows operations on values to return early in the case of errors.

While I am sensitive to the desire to cast the array once, there are a couple of complexities that I think you're missing in this proposal:
- the scalar functions are operating on row data: they're _always_ going to be iterating over the array row-by-row. We can move where the cast is, maybe save a few cylces, but the margin is quite slim.
- the return types of an operation may not be the same as the source type of the function.

Reply:
> - handling columns of arbitrary types is hard, and so if we don't allow ourselves this kind of operation, we'll end up having functions that are limited to only some types (which is more friction for users), or handling the same type conversions in different places.
> - if we don't allow ourselves to convert columns to scalars in functions writing functions for GlareDB (of any kind) becomes much more difficult and our time to delivery goes down.

Usually i'd agree with this sentiment that we should write the code in a way that makes it easier to develop, but once again, this is a performance critical path so the normal rules no longer apply. If we can introduce zero cost, or near zero cost abstractions, by all means lets do it. But the abstractions in this PR don't justify their cost.

Reply:
> While I am sensitive to the desire to cast the array once, there are a couple of complexities that I think you're missing in this proposal:

> - the scalar functions are operating on row data: they're always going to be iterating over the array row-by-row. We can move where the cast is, maybe save a few cylces, but the margin is quite slim.


While yes, we need to operate on the individual elements (row-by-row), downcasting an `ArrayRef<T>` -> `StringArray` is not the same as casting `T` -> `ScalarValue<T>` inside the loop. There is no reason we need to operate on `ScalarValue` instead of just `T`.

ArrayRef is essentially a type erased version of an array, so it's incredibly cheap to downcast it to a concrete type. (just a vtable lookup), while the `ScalarValue::try_from_array` is doing a vtable lookup `n` times instead of once. 

You are also allocating a buffer to transfer from `ArrayRef` to `Vec<T>` when we can just go directly from `ArrayRef` to `Vec<T>` without any iterations instead of the two iterations done in the PR.



```rust
for idx in 0..arr.len() { // <-- iterates entire array
    values.push(op(ScalarValue::try_from_array(arr, idx)?)?);
}

Ok(ColumnarValue::Array(ScalarValue::iter_to_array(
    values.into_iter(), // <-- iterates entire array
)?))
```

Reply:
> - the return types of an operation may not be the same as the source type of the function.

I don't understand how that is relevant? If the signature is `|Utf8 | LargeUtf8| -> Utf8`, theres only 2 brances so we can easily handle that by checking the dtype and handling the concrete arrays individually. 

> I think your example is an interesting start, but I'd really like to see a proposal that allows operations on values to return early in the case of errors.

Maybe it was a poor example, but there's nothing stopping us from handling errors however we want by operating on `ArrayRef`. There should be no disparity in what we can do by using ArrayRef